test/common/upstream/load_balancer_impl_test.cc

```diff
@@ -2880,6 +2880,117 @@ TEST_P(LeastRequestLoadBalancerTest, PNC) {
+                              makeTestHost(info_, "tcp://127.0.0.1:82", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:83", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:84", simTime())};
+  hostSet().hosts_ = hostSet().healthy_hosts_;
+  hostSet().runCallbacks({}, {}); // Trigger callbacks. The added/removed lists are not relevant.
+
+  hostSet().healthy_hosts_[0]->stats().rq_active_.set(3);
+  hostSet().healthy_hosts_[1]->stats().rq_active_.set(3);
+  hostSet().healthy_hosts_[2]->stats().rq_active_.set(1);
+  hostSet().healthy_hosts_[3]->stats().rq_active_.set(1);
+  hostSet().healthy_hosts_[4]->stats().rq_active_.set(1);
+
+  // Creating various load balancer objects with different choice configs.
+  envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest lr_lb_config;
+  lr_lb_config.mutable_choice_count()->set_value(2);
+  // Enable full table scan on hosts.
+  lr_lb_config.set_selection_method(
+    envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest::FULL_SCAN);
+  common_config_.mutable_healthy_panic_threshold()->set_value(0);
+
+  LeastRequestLoadBalancer lb{priority_set_, nullptr, stats_,       runtime_,
+                              random_,       1,       lr_lb_config, simTime()};
+
+  // Indices 2-4 are tied for least.
+  // Random numbers are generated whenever a tie is encountered, which will occur at:
+  // - Index 1 (tied with index 0)
+  // - Index 3 (tied with index 2) -> use 998 so that 998 % 2 ties == 0 to select index 3
+  // - Index 4 (tied with indices 2-3) -> use 4 so that 4 % 3 ties != 0 to keep index 3
+  EXPECT_CALL(random_, random())
+    .WillOnce(Return(0))
+    .WillOnce(Return(0))
+    .WillOnce(Return(998))
+    .WillOnce(Return(4));
+  EXPECT_EQ(hostSet().healthy_hosts_[3], lb.chooseHost(nullptr));
+
+  // Indices 2-4 are tied for least.
+  // Random numbers are generated whenever a tie is encountered, which will occur at:
+  // - Index 1 (tied with index 0)
+  // - Index 3 (tied with index 2) -> use 998 so that 998 % 2 ties == 0 to select index 3
+  // - Index 4 (tied with indices 2-3) -> use 6 so that 6 % 3 ties == 0 to select index 4
+  EXPECT_CALL(random_, random())
+    .WillOnce(Return(0))
+    .WillOnce(Return(0))
+    .WillOnce(Return(998))
+    .WillOnce(Return(6));
+  EXPECT_EQ(hostSet().healthy_hosts_[4], lb.chooseHost(nullptr));
+
+  // Indices 2-4 are tied for least.
+  // Random numbers are generated whenever a tie is encountered, which will occur at:
+  // - Index 1 (tied with index 0)
+  // - Index 3 (tied with index 2) -> use 999 so that 998 % 2 ties != 0 to keep index 2
+  // - Index 4 (tied with indices 2-3) -> use 4 so that 4 % 3 ties != 0 to keep index 2
+  EXPECT_CALL(random_, random())
+    .WillOnce(Return(0))
+    .WillOnce(Return(0))
+    .WillOnce(Return(999))
+    .WillOnce(Return(4));
+  EXPECT_EQ(hostSet().healthy_hosts_[2], lb.chooseHost(nullptr));
```

Comment:
It may be better (and potentially less fragile) to just use a genuine `random_` generator and just call `chooseHost()` like 1e6 times. Then you can verify that each host was picked the expected fraction of the time. I believe you do that via `EXPECT_NEAR(...)`.

I just worry about changes to `chooseHost` requiring changes to this test every time and frustrating folks down the line.

Reply:
Agreed. Just pushed a commit with this change (and the others).

It was straightforward, other than figuring out why `#include "source/common/common/random_generator.h"` wasn't recognized (needed to modify a BUILD file).